applications/ConstitutiveLawsApplication/custom_constitutive/masonry_orthotropic_damage_plane_stress_2d.cpp

```diff
@@ -0,0 +1,926 @@
+            angle_to_damage, projected_material);
+
+        // compute the equivalent stress measures
+        this->CalculateEquivalentStressTension(data, projected_material, effective_stress_mapped_isotropic, principal_stresses_mapped_isotropic, mUniaxialStressTension);
+        this->CalculateEquivalentStressCompression(data, projected_material, effective_stress_mapped_isotropic, principal_stresses_mapped_isotropic, mUniaxialStressCompression);
+
+        if (mUniaxialStressTension > mThresholdTension)
+            mThresholdTension = mUniaxialStressTension;
+        this->CalculateDamageTension(projected_material, mThresholdTension, mDamageTension);
+
+        if (mUniaxialStressCompression > mThresholdCompression)
+            mThresholdCompression = mUniaxialStressCompression;
+        this->CalculateDamageCompression(projected_material, mThresholdCompression, mDamageCompression);
+
+        mCurrentThresholdTension = mThresholdTension;
+        mCurrentThresholdCompression = mThresholdCompression;
+
+        //mDamageTension = std::max(mDamageTension, mDamageCompression);
+
+        // Compute the stresses to isotropic space
+        array_1d<double, 3> total_tension_stress_mapped_isotropic = (1.0 - mDamageTension) * effective_tension_stress_mapped_isotropic;
+        array_1d<double, 3> total_compression_stress_mapped_isotropic = (1.0 - mDamageCompression) * effective_compression_stress_mapped_isotropic;
+
+        // Back rotation of final Predictive Stress
+        noalias(rPredictiveStressVector) = prod(transformation_matrices.InverseTransformationMatrixTension, total_tension_stress_mapped_isotropic);
+        noalias(rPredictiveStressVector) += prod(transformation_matrices.InverseTransformationMatrixCompression, total_compression_stress_mapped_isotropic);
+
+    }
+    /***********************************************************************************/
+    /***********************************************************************************/
+    void MasonryOrthotropicDamagePlaneStress2DLaw::CheckDamageLoadingUnloading(
+        bool& is_damaging_tension,
+        bool& is_damaging_compression)
+    {
+        const double F_tension = mUniaxialStressTension - mCurrentThresholdTension;
+        const double F_compression = mUniaxialStressCompression - mCurrentThresholdCompression;
+
+        is_damaging_tension = (F_tension > 0.0)
+            ? true : false;
+        is_damaging_compression = (F_compression > 0.0)
+            ? true : false;
+    }
+    /***********************************************************************************/
+    /***********************************************************************************/
+    void MasonryOrthotropicDamagePlaneStress2DLaw::CalculateTangentTensor(
+        Parameters& rValues,
+        const Matrix& rElasticityMatrix,
+        const array_1d<double, 3>& rStrainVector,
+        const array_1d<double, 3>& rPredictiveStressVector,
+        CalculationData& data,
```

Comment:
We have implemented and tested a numerical perturbation methodology! maybe you could use the same as ours :)